### ISSUE TITLE
Retweet isn't working!

### ISSUE THREAD

**Issue by crealol** [Rishikant181/Rettiwt-API#639]
Hello! First of all thank you for all the work you are doing!

I think i stumbled across a bug:

Retweet isn't working, but giving "true" back. 
Everything else is working perfectly.

I'm using a proxy and also tried with fetcherservice, but with the same result > isnt retweeting but there is no error.

**Comment by Rishikant181** [comment]
Okay I'm checking

**Comment by Rishikant181** [comment]
It's a strange issue indeed. I checked the response and the response says OK. Twitter is even returning an ID allocated to the retweet, but when cross-referenced, no retweet with that ID exists. I'll investigate it further and keep you updated.

**Comment by Rishikant181** [comment]
@crealol For reasons unknown, it's working again now. I think it was an issue from Twitter's side. Let me know if it works for you too.

**Comment by gratatouilley** [comment]
I'm suffering same issue for few days back for sending and reply tweet.

**Comment by Rishikant181** [comment]
@gratatouilley I tried it approximately 6 hrs back from now, and then, it didn't work, although the raw response confirmed that it did retweet and also provided a retweet id, but in reality, no retweet was made.

Now, it's working as expected.

Can you confirm if it's working (or not) for you now?

**Comment by gratatouilley** [comment]
@Rishikant181 tried manual trigger seems working perfectly. but here is the automatic one that happened just about less than 15 mins ago.

```1|notification  | [Rettiwt-API] [2024-11-02T19:02:47.519Z] [AUTHORIZATION] {"authenticated":true}
1|notification  | [Rettiwt-API] [2024-11-02T19:02:47.519Z] [VALIDATE] {"target":"POST_ARGS"}
1|notification  | [Rettiwt-API] [2024-11-02T19:02:47.522Z] [GET] {"target":"HTTPS_PROXY_AGENT"}
1|notification  | [Rettiwt-API] [2024-11-02T19:02:47.522Z] [GET] {"target":"USER_CREDENTIAL"}
1|notification  | 3 November 2024 Jam 02:02:51 WIB: info: [a37a1cd4-dd33-4ad7-a8c2-700fc683aafb] Rettiwt operation succeeded on attempt 1
1|notification  | 3 November 2024 Jam 02:02:51 WIB: info: [a37a1cd4-dd33-4ad7-a8c2-700fc683aafb] TweetID 1852760490038178287 replied with media using Rettiwt 1852788478653985081
1|notification  | replyTweetWithMedia-a37a1cd4-dd33-4ad7-a8c2-700fc683aafb: 6.718s
```

It said succeed but the tweet isn't exist. Then it will reply the sent tweet (log below)
```1|notification  | 3 November 2024 Jam 02:03:01 WIB: info: [0f5dc49f-6078-42df-aa89-382013b47022] Attempt 1/30 for Rettiwt operation
1|notification  | [Rettiwt-API] [2024-11-02T19:03:01.488Z] [REQUEST] {"resource":"TWEET_POST","args":{"tweet":{"text":"Censored","replyTo":"1852788478653985081","media":[{"id":"1852788518504062976"},{"id":"1852788518751485952"},{"id":"1852788518663356416"},{"id":"1852788518596268032"}]}}}
1|notification  | [Rettiwt-API] [2024-11-02T19:03:01.489Z] [AUTHORIZATION] {"authenticated":true}
1|notification  | [Rettiwt-API] [2024-11-02T19:03:01.489Z] [VALIDATE] {"target":"POST_ARGS"}
1|notification  | [Rettiwt-API] [2024-11-02T19:03:01.490Z] [GET] {"target":"HTTPS_PROXY_AGENT"}
1|notification  | [Rettiwt-API] [2024-11-02T19:03:01.490Z] [GET] {"target":"USER_CREDENTIAL"}
1|notification  | 3 November 2024 Jam 02:03:05 WIB: warn: [0f5dc49f-6078-42df-aa89-382013b47022] Rettiwt attempt 1 failed: {}
1|notification  | 3 November 2024 Jam 02:03:07 WIB: info: [0f5dc49f-6078-42df-aa89-382013b47022] Attempt 2/30 for Rettiwt operation
1|notification  | [Rettiwt-API] [2024-11-02T19:03:07.202Z] [REQUEST] {"resource":"TWEET_POST","args":{"tweet":{"text":"Censored","replyTo":"1852788478653985081","media":[{"id":"1852788518504062976"},{"id":"1852788518751485952"},{"id":"1852788518663356416"},{"id":"1852788518596268032"}]}}}
1|notification  | [Rettiwt-API] [2024-11-02T19:03:07.202Z] [AUTHORIZATION] {"authenticated":true}
1|notification  | [Rettiwt-API] [2024-11-02T19:03:07.202Z] [VALIDATE] {"target":"POST_ARGS"}
1|notification  | [Rettiwt-API] [2024-11-02T19:03:07.203Z] [GET] {"target":"HTTPS_PROXY_AGENT"}
1|notification  | [Rettiwt-API] [2024-11-02T19:03:07.203Z] [GET] {"target":"USER_CREDENTIAL"}
1|notification  | 3 November 2024 Jam 02:03:11 WIB: warn: [0f5dc49f-6078-42df-aa89-382013b47022] Rettiwt attempt 2 failed: {}
```

There is no error message too.

**Comment by Rishikant181** [comment]
@gratatouilley Now I got hit with `You have reached you daily limit for posting Tweets.....` on my alt account, even though I can retweet from the web app. Definitely something is awry. I'm investigating it further.

**Comment by jonasxk** [comment]
Retweeting for me works with text and video-based content. It seems the only problems seem to be with the images?

**Comment by Rishikant181** [comment]
Found the issue, but right now, I don't have a solution. 

The absence of the header `x-client-transaction-id` is causing the issue. Earlier, no such header was required. I don't know how that ID is generated of where it's derived from. 

Any input is welcome.

**Comment by sahas-si** [comment]
This could be of help, someone has created the twitter tid generator, but in golang -> https://github.com/yeyuchen198/twitter-tid-generator?tab=readme-ov-file

**Comment by Rishikant181** [comment]
@sahas-si Thanks for this. I'll try to port it into TS, but I'm finding it difficult to find some time from my schedule.

**Comment by gratatouilley** [comment]
@Rishikant181 any updates?

**Comment by Rishikant181** [comment]
@gratatouilley No, unfortunately. Having difficulty trying to squeeze out time for this. I apologize for the inconvenience.

**Comment by gratatouilley** [comment]
@Rishikant181 No worries, take your time brother. Happy weekend! ^^

**Comment by Asukaz** [comment]
just tried reply test, the first time is good, then it not success, like your said, it can get reply tweet id, but actually it is empty, just curious why the first time is good.

**Comment by ronaldlangeveld** [comment]
Can confirm, haven't been able to get the `retweetedTweet` entity data.

**Comment by acid-drop** [comment]
@Rishikant181 I am also noticing this issue with posting tweets. Ocassionally they post, ocassionally it returns an id that doesn't really exists, sometimes it returns "undefined". When posting a tweet via the web I see that `x-client-transaction-id` is present. Thanks for all your work on this!

**Comment by Rishikant181** [comment]
@acid-drop Yeah generating that `x-client-transaction-id` has been an issue. Once we implement it's generation in the code, it should fix this (as well as other posting-related issues).

**Comment by kingkej** [comment]
> [@acid-drop](https://github.com/acid-drop) Yeah generating that `x-client-transaction-id` has been an issue. Once we implement it's generation in the code, it should fix this (as well as other posting-related issues).

I can try implementing it, but seems like https://github.com/yeyuchen198/twitter-tid-generator is not working anymore. I also tried https://github.com/iSarabjitDhiman/TweeterPy/blob/master/tweeterpy/tid/transaction.py, but it has not immediate effect on retweeting. Also seems like "likes" are not working either, sometimes you will get a successful like response, but it won't appear actually. 
So now I am thinking of digging into Python script and give it one more try.

**Comment by acid-drop** [comment]
> I can try implementing it, but seems like https://github.com/yeyuchen198/twitter-tid-generator is not working anymore. I also tried https://github.com/iSarabjitDhiman/TweeterPy/blob/master/tweeterpy/tid/transaction.py, but it has not immediate effect on retweeting. Also seems like "likes" are not working either, sometimes you will get a successful like response, but it won't appear actually. So now I am thinking of digging into Python script and give it one more try.

Maybe https://github.com/yeyuchen198/twitter-tid-generator isn't working any more because twitter changed the hash salt from "bird" to "obfiowerehiring" (interestingly!) .. but the python port has the correct salt and seems easier to follow anyway imho.

**Comment by Ansaal** [comment]
Any update on this? Anything I can help with?

**Comment by Rishikant181** [comment]
@Ansaal [This](https://github.com/iSarabjitDhiman/TweeterPy/blob/6f127c7095bf4be583728f191ce6d42de0461628/tweeterpy/tid/transaction.py#L123) seems to be the solution, but I'm finding hard to find some time for porting this over to TypeScript.

**Comment by Ansaal** [comment]
I can port that function over. I have no idea where to integrate it though. Will create a PR with it.

**Comment by Rishikant181** [comment]
@Ansaal Just put it in `helper/tid` directory.

**Comment by Ansaal** [comment]
https://github.com/Rishikant181/Rettiwt-API/pull/663
